lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs

```diff
@@ -0,0 +1,221 @@
+    /// of times this metric has been updated. If the metric is immutable, it keeps a record of updates.
+    /// Once the data has been processed, the records and count will reset.
+    /// </summary>
+    [JsonObject]
+    public sealed class MetricTracker : IObserver<IMetric>
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(MetricsData));
+
+        private bool _keepUpdateHistory;
+
+        private IDisposable _unsubscriber;
+
+        [JsonProperty]
+        private IMetric _mirror;
+
+        /// <summary>
+        /// List of the history of values this metric has held. If _keepUpdateHistory is false, only holds current value.
+        /// </summary>
+        [JsonProperty]
+        private ConcurrentQueue<MetricRecord> _records;
+        
+        /// <summary>
+        /// Number of times metric has been updated since last processed.
+        /// </summary>
+        [JsonProperty]
+        internal int ChangesSinceLastSink;
+
+        /// <summary>
+        /// Constructor for metricData
+        /// </summary>
+        /// <param name="metric"></param>
+        /// <param name="initialValue"></param>
+        internal MetricTracker(IMetric metric)
+        {
+            Subscribe(metric);
+            _mirror = metric;
+            ChangesSinceLastSink = 0;
+            _keepUpdateHistory = metric.IsImmutable;
+            _records = new ConcurrentQueue<MetricRecord>();
+            _records.Enqueue(CreateMetricRecord(metric));
+        }
+
+        [JsonConstructor]
+        internal MetricTracker(IMetric _mirRor, int changesSinceLastSink, ConcurrentQueue<MetricRecord> _recorDs)
+        {
+            _mirror = _mirRor;
+            _records = new ConcurrentQueue<MetricRecord>(_recorDs);
+            ChangesSinceLastSink = changesSinceLastSink;
+        }
+
+        private MetricTracker(IMetric mirror, ConcurrentQueue<MetricRecord> records, int changes, bool history)
+        {
+            _mirror = mirror;
+            _records = records;
+            ChangesSinceLastSink = changes;
+            _keepUpdateHistory = history;
+        }
+
+        /// <summary>
+        /// Flush records.
+        /// </summary>
+        internal ConcurrentQueue<MetricRecord> FlushChangesSinceLastSink()
+        {
+            ConcurrentQueue<MetricRecord> records = new ConcurrentQueue<MetricRecord>();
+            MetricRecord record;
+            while (_records.TryDequeue(out record))
+            {
+                records.Enqueue(record);
+            }
+            ChangesSinceLastSink = 0;
+            return records;
+        }
+
+        /// <summary>
+        /// When new metric data is received, update the value and records so it reflects the new data.
+        /// </summary>
+        /// <param name="metric">Metric data received.</param>
+        internal MetricTracker UpdateMetric(MetricTracker metric)
+        {
+            _mirror = metric.GetMetric();
+            if (metric.ChangesSinceLastSink > 0)
```

Comment:
This doesnt look to be thread-safe - if two threads came here and checked if metric.ChangesSinceLastSink > 0, they would both hit the last code of incrementing ChangesSinceLastSink. So you would miss an event if history were being tracked.